src/repositories/ContractEntityRepository.ts

```diff
@@ -0,0 +1,175 @@
+ */
+export enum EntityType {
+  StakeRequesteds = 'stakerequesteds',
+  StakeIntentDeclareds = 'stakeintentdeclared',
+  StateRootAvailables = 'staterootavailable',
+  GatewayProvens = 'gatewayproven',
+  StakeIntentConfirmeds = 'stateintentconfirmed',
+  StakeProgresseds = 'stakeprogressed',
+  MintProgresseds = 'mintprogressed',
+}
+
+/**
+ * Stores instances of ContractEntity.
+ *
+ * Class enables creation, update and retrieval of ContractEntity objects.
+ * On construction it initializes underlying database model.
+ */
+export class ContractEntityRepository extends Subject<ContractEntity> {
+  /* Public Functions */
+
+  public constructor(initOptions: InitOptions) {
+    super();
+    ContractEntityModel.init(
+      {
+        contractAddress: {
+          type: DataTypes.STRING,
+          allowNull: false,
+          validate: {
+            isAlphanumeric: true,
+            len: [42, 42],
+          },
+          primaryKey: true,
+        },
+        entityType: {
+          type: DataTypes.ENUM({
+            values: [
+              EntityType.StakeIntentDeclareds,
+              EntityType.StakeRequesteds,
+              EntityType.StateRootAvailables,
+              EntityType.StakeIntentConfirmeds,
+              EntityType.StakeProgresseds,
+              EntityType.MintProgresseds,
+              EntityType.GatewayProvens,
+            ],
+          }),
+          primaryKey: true,
+        },
+        timestamp: {
+          type: DataTypes.BIGINT,
+          allowNull: false,
+        },
+      },
+      {
+        ...initOptions,
+        modelName: 'ContractEntityModel',
+        tableName: 'contract_entities',
+      },
+    );
+  }
+
+  /**
+   * Saves a contract entity model in the repository.
+   * If a contract entity does not exist, it creates, otherwise updates.
+   *
+   * Function ignores (does not set to null) undefined (optional) fields
+   * from the passed stake request object.
+   *
+   * @param contractEntity Contract entity object to update.
+   *
+   * @returns Newly created or updated contract entity object.
+   */
+  public async save(contractEntity: ContractEntity): Promise<ContractEntity | null> {
+    const definedOwnProps: string[] = Utils.getDefinedOwnProps(contractEntity);
+
+    ContractEntityModel.upsert(
+      contractEntity,
+      {
+        fields: definedOwnProps,
+      },
+    );
+
+    const updatedContractEntity = await this.get(contractEntity.contractAddress, contractEntity.entityType);
+    this.newUpdate(updatedContractEntity as ContractEntity);
+
+    return updatedContractEntity;
+  }
+
+  /**
+   * Fetches ContractEntity data from database if found. Otherwise returns null.
+   * @param contractAddress Address of the contract.
+   * @param entityType Type of the entity.
+   * @returns ContractEntity object containing values which satisfy the `where` condition.
+   */
+  public async get(
+    contractAddress?: string,
```

Comment:
why "?" 🤔 
contractAdress and entityType needs to be mandatory here.

Reply:
Changed it to mandatory.